### ISSUE TITLE
API Proposal: Type.IsAssignableTo(Type)

### ISSUE THREAD

**Issue by benaadams** [dotnet/corefx#40666]
The api `Type.IsAssignableFrom(Type)` is [quite common to use backwards](https://twitter.com/Nick_Craver/status/1166820786818179072) because it flows unexpectedly with surrounding code (swapping the subject and object; the type you are interested in becoming a parameter rather than staying as the callee).

To alleviate this introducing a reversal of the parameters either as extension or method would be helpful to differentiate and provide clarity:

```csharp
partial class Type
{
    public bool Type.IsAssignableTo(Type type) 
        => type.IsAssignableFrom(this);
}
```
or
```csharp
partial static class TypeExtensions
{
    public static bool Type.IsAssignableTo(this Type type, Type assignableType)
        => assignableType.IsAssignableFrom(type);
}
```

/cc @NickCraver @davkean

**Comment by stevedesmond-ca** [comment]
Depending on what the goal/intent of this method is, I really liked @NickCraver's method name [suggestion](https://twitter.com/Nick_Craver/status/1166822168791015425?s=19) `IsBasedOn()`, or maybe `IsSubtypeOf()` to go along with `IsSubclassOf()`?

**Comment by stevedesmond-ca** [comment]
Or, alternatively...
![why not both meme](https://media.tenor.com/images/a8cc44455fc5c8526af8125c70b47ff7/tenor.gif)

**Comment by benaadams** [comment]
While "based on" and/or "subtype" work with a class hierarchy; don't think they convey the correct relationship with an interface?

**Comment by vcsjones** [comment]
> don't think they convey the correct relationship with an interface?

I think this is part of what makes `IsAssignableFrom` quirky for me. The method name attempts to be so precise in what it does ("assignable") that it causes me to stop and think about it every time and has lost meaning. This API name is "technically correct, the best kind of correct" but harder for me to immediately understand the relationship between the call site and the parameter.

**Comment by benaadams** [comment]
> The method name attempts to be so precise in what it does ("assignable") that it causes me to stop and think about it every time and has lost meaning.

Its also incorrect as it doesn't take into account explicit and implicit casting which are technically assignable (in C#)

**Comment by oocx** [comment]
"IsAssignableFrom" is an api that I always get wrong on the first try. I'm not sure if "IsAssignableTo" would work better for me, but I like "IsSubtypeOf", I think that would be intuitive to use.

**Comment by stevedesmond-ca** [comment]
I think part of getting this right is figuring out what questions devs are trying to ask: e.g.
- Can I use X as a plugin for type Y?
- Can I safely cast X to Y?
- Does X implement Y?
While the answers may be the same for all of these, the questions are subtly different.

What are some other questions we're currently currently using `IsAssignableFrom()` to answer?

**Comment by reflectronic** [comment]
> I like "IsSubtypeOf", I think that would be intuitive to use.

I would be hesitant to do that considering there's already an [`IsSubclassOf`](https://docs.microsoft.com/en-us/dotnet/api/system.type.issubclassof?view=netcore-3.0) with slightly different semantics (it ignores interfaces).

**Comment by stevedesmond-ca** [comment]
Other ideas for names / extensions I've created in the past: `Implements(otherType)`, `InheritsFrom(otherType)` -- again, I think it depends on the user's goals and how directly the framework should support those :thinking:

**Comment by GSPP** [comment]
From that [Twitter thread](https://twitter.com/lucasmeijer/status/1166822122725040130):

> it's the usb stick of the dotnet world.

True :rofl: 

<hr>

But is `IsAssignableTo` actually clearer? I think it has the same source of confusion. And now you additionally need to pick between the "from" and "to" APIs.

I think the problem is not the from/to direction. The problem is that "assignable" is not intuitively understood. If this usability issue is to be fixed it should be by changing the verb, not the to/from suffix. When doing that we can also change the direction if that seems desirable. But that would be a secondary thing.

**Comment by zezba9000** [comment]
I could not agree more with "IsAssignableTo" or "IsClassOrSubclassOf" as it flows with the rest of my logic more normally.

**Comment by tamlin-mike** [comment]
> Its also incorrect as it doesn't take into account explicit and implicit casting

Indeed. How such a misleading name with such a broken implementation got to be placed in public API's is baffling.

I've been banging my head against this for the last few days. I have two types - one data struct and one (generic) XML serializer. `IsAssignableFrom` is, based on the name, _completely broken_. It responds `false` for both ways of assigning between the two, yet when actually assigning between the two types it's not only compilable but also works as expected. All this pain due to the bad naming and/or inadequate implementation not accounting for `implicit` conversion operators or even (conversion) contructors.

Since this method is evidently not about assignability, merely the limited subset of assignability that comes with derivation, **the method shouldn't even have "Assignable" in the name**.

**Comment by GSPP** [comment]
@tamlin-mike I think your position is that either the reflection APIs should implement C# semantics cleanly or they should not suggest that they do.

I think the APIs are indeed a weird hybrid between C#/VB.NET and CLR semantics. The whole `Binder` infrastructure, for example, is a pretty weird design mistake.

It would have been the cleanest solution if the reflection APIs were implementing purely CLR semantics and giving raw access to data. The C# semantics should have been layered on top of that with another optional library.

Now, we are stuck with this forever for compatibility.

**Comment by jkotas** [comment]
> the APIs are indeed a weird hybrid between C#/VB.NET and CLR semantics. The whole Binder infrastructure, for example, is a pretty weird design mistake.

Yes, the Binder infrastructure is a weird hybrid.

`Type.IsAssignableTo` that this issue is about implements `compatible-with` CLR semantics as defined in ECMA-335. I agree that the name could have been better, but the behavior is not a hybrid.

**Comment by GSPP** [comment]
> but the behavior is not a hybrid

That is interesting to learn and makes much sense.
